pkg/R/RDD.R

```diff
@@ -1323,6 +1323,88 @@ setMethod("zipWithIndex",
            lapplyPartitionsWithIndex(x, partitionFunc)
          })
 
+#' Zip an RDD with another RDD.
+#'
+#' Zips this RDD with another one, returning key-value pairs with the
+#' first element in each RDD second element in each RDD, etc. Assumes
+#' that the two RDDs have the same number of partitions and the same
+#' number of elements in each partition (e.g. one was made through
+#' a map on the other).
+#'
+#' @param x An RDD to be zipped.
+#' @param other Another RDD to be zipped.
+#' @return An RDD zipped from the two RDDs.
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd1 <- parallelize(sc, 0:4)
+#' rdd2 <- parallelize(sc, 1000:1004)
+#' collect(zipRDD(rdd1, rdd2))
+#' # list(list(0, 1000), list(1, 1001), list(2, 1002), list(3, 1003), list(4, 1004))
+#'}
+#' @rdname zipRDD
+#' @aliases zipRDD,RDD
+setMethod("zipRDD",
+          signature(x = "RDD", other = "RDD"),
+          function(x, other) {
+            n1 <- numPartitions(x)
+            n2 <- numPartitions(other)
+            if (n1 != n2) {
+              stop("Can only zip RDDs which have the same number of partitions.")
+            }
+
+            # Check if corresponding partitions of both RDDs have the same
+            # number of elements.
+            getPartitionLengthFunc <- function(part) {
+              list(length(part))
+            }
+            partitionLengths1 <- collect(lapplyPartition(x, getPartitionLengthFunc))
```

Comment:
Hmm - so one thing about this approach is that it will trigger a Spark job on both the RDDs. Could we do something in the task like what is done in Scala version ? [1] i.e. inside the map function we can check if the two lengths are equal or exit with `stop(unequal partitions)` ?

[1] https://github.com/apache/spark/blob/c49d156624624a719c0d1262a58933ea3e346963/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L746